locallib.php

```diff
@@ -623,6 +623,88 @@ private function process_action_manual_allocation() {
         return $output;
     }
 
+    /**
+     * Retrieve all used groups in rateable choices.
+     *
+     * @return array of group ids used in rateable choices
+     */
+    public function get_all_groups_of_choices(): array {
+        $rateablechoiceswithgrouprestrictions = array_filter($this->get_rateable_choices(),
+                fn($choice) => !empty($choice->usegroups) && !empty($this->get_choice_groups($choice->id)));
+        $rateablechoiceids = array_map(fn($choice) => $choice->id, $rateablechoiceswithgrouprestrictions);
+        $groupids = [];
+        foreach ($rateablechoiceids as $choiceid) {
+            $groupids = array_merge($groupids, array_map(fn($group) => $group->id, $this->get_choice_groups($choiceid)));
+        }
+        return array_unique($groupids);
+    }
+
+    /**
+     * Helper method returning an array of groupids belonging to the groups the user is member in.
+     *
+     * If the user is not a member of any group an empty array is being returned. Only group ids of groups defined in the
+     * choices restrictions are being considered here.
+     *
+     * @param int $userid the id of the user we want to get the group ids he/she belongs to
+     * @return array of group ids the user belongs to, not including groups which are not specified in at least one of the choices'
+     *  group restrictions
+     */
+    public function get_user_groupids(int $userid): array {
+        $groups = groups_get_user_groups($this->ratingallocate->course, $userid)[0];
+        if (empty($groups)) {
+            return [];
+        } else {
+            return array_filter($groups, fn($group) => in_array($group, $this->get_all_groups_of_choices()));
```

Comment:
It seems to work also without if statement, if you call array_filter on an empty array the result is the empty array.

Reply:
IIRC I added this to prevent the issue of `$groups` not being defined

Reply:
fair enough. Using `$groups  = groups_get_user_groups(...) ?? []`is barely an improvement